### PR TITLE
Facebook, Twitter and Instagram shareSingle - Fix on iOS 11

### PR DESCRIPTION
**Pull request by IbrahimSulai**
* Fix for Facebook and Twitter shareSingle is not working on iOS 11.
* Handled the image share option on Instagram.

### CODE REVIEW

**Comment by IbrahimSulai** [comment]
The above fix will solve the below issues:

https://github.com/react-native-community/react-native-share/issues/232

https://github.com/react-native-community/react-native-share/issues/174

https://github.com/react-native-community/react-native-share/issues/389

https://github.com/react-native-community/react-native-share/issues/388

**Comment by MateusAndrade** [comment]
WOW! This looks really good! Im running out of time this week to review this. But doing a overview its LGTM! 👍 

@mikehardy you can give a look at this?

**Comment by mikehardy** [comment]
I love the idea of fixing a pile of bugs for people, and this looks like good code - I read through it - the problem (for me, it’s my problem) is that I’m not an iOS developer, I can only pretend I am because of react-native. I’m your Android / Java review guy :-). I wish I could look at this and say “yes!” but I just don’t have the skills.

What I do have is some basic quality control questions though - @IbrahimSulai can you help me out?

1. react-native supports back to ios9 through current (12.x) currently, and our module is in harmony with that. Have you tested these changes on ios9, 10 and 12? I assume 11 yes since it’s in the PR name…
1. Are you using this in production (or is it in testing for release) in an app? How many users does the app have, and does it spread across those ios levels?
1. Can you handle any follow-on issues with this specific PR if it causes problems? I frequently propose a PR in a repo and it’s 99% and has one crash bug for instance, and I’ll chase my first PR with a fix, then all is quiet. Will you listen for problems for a bit to make sure it lands well?

I ask those things because even in situations where I don’t personally have the skills, if all those answers are a “yes”, I am biased towards giving it a shot. Sometimes things break, but we all move forward, and it’s awesome to get PRs with fixes :-)

**Comment by IbrahimSulai** [comment]
> I love the idea of fixing a pile of bugs for people, and this looks like good code - I read through it - the problem (for me, it’s my problem) is that I’m not an iOS developer, I can only pretend I am because of react-native. I’m your Android / Java review guy :-). I wish I could look at this and say “yes!” but I just don’t have the skills.

> 
> What I do have is some basic quality control questions though - @IbrahimSulai can you help me out?
> 
> 1. react-native supports back to ios9 through current (12.x) currently, and our module is in harmony with that. Have you tested these changes on ios9, 10 and 12? I assume 11 yes since it’s in the PR name…

> [@IbrahimSulai ] **I have tested in iOS 11 and iOS 12  since 96% of iOS devices are running in iOS 11 and above. Also, I don't have a lower version device with me for testing as of now. Expecting these changes won't break anything in lower versions and ready to take up any issues related to lower versions.**

> 2. Are you using this in production (or is it in testing for release) in an app? How many users does the app have, and does it spread across those ios levels?

> [@IbrahimSulai ] **Yes, our application is in production and the changes are in the beta release. The minimum version of the app supports is iOS 11.**

> 3. Can you handle any follow-on issues with this specific PR if it causes problems? I frequently propose a PR in a repo and it’s 99% and has one crash bug for instance, and I’ll chase my first PR with a fix, then all is quiet. Will you listen for problems for a bit to make sure it lands well?

>[@IbrahimSulai ] **Sure @mikehardy , I will be continuously available and support the RN community especially this share-component.**

> I ask those things because even in situations where I don’t personally have the skills, if all those answers are a “yes”, I am biased towards giving it a shot. Sometimes things break, but we all move forward, and it’s awesome to get PRs with fixes :-)

**Comment by IbrahimSulai** [comment]
@mikehardy - I have answered the questions above in inline. Please review and merge the same. Thanks

**Comment by mikehardy** [comment]
I’m not satisifed by the “have not tested below 11” and “our app doesn’t use 11” answers - these items are subtle, it needs testing on 9 and 10 to make sure, you don’t need lower level devices I believe this stuff is simulator capable and you can run iOS 9 and iOS 10 simulators yes? Knowing that the code itself is in production sounds great though, that means for your intended audience (11+) you know the code works and we can have more confidence in it.

And very glad of course that you’ll listen in case something goes wrong.

If someone (yourself or perhaps someone on the linked issues) could affirm it works on ios 9 and 10 then we may be ready

**Comment by mikehardy** [comment]
(one reason I stick on testing the actual APIs when testing something version specific - notifications for instance is massively different between 9 and 10 - sure, not many people use it, but a test is easier than responding to bug reports from the few that *must* have it for 9)

**Comment by IbrahimSulai** [comment]
@mikehardy Yes, that makes sense to me. I don't have iOS 9 devices with me at present. I will try to get one for testing. Any help from the community for testing on iOS 9 is appreciated.

**Comment by mikehardy** [comment]
@IbrahimSulai the simulator should suffice? You can still get ios 9 and 10 images for the simulator (I use them for my app)

**Comment by jgcmarins** [comment]
Awesome work guys.
I don't see any problem merging this one.
Most Apple users are at iOS 11 or above: https://developer.apple.com/support/app-store/

**Comment by IbrahimSulai** [comment]
@mikehardy @MateusAndrade 
I have tested the implemented code in the simulator iOS 9 and iOS 10. It's working fine as expected and not breaking the existing functionality.

For testing purpose, I have created the below repository, which includes my current implementation also. Please have a look and try.
https://github.com/IbrahimSulai/react-native-share-test/tree/master

**Comment by mikehardy** [comment]
@jgcmarins my app has many ios 10 and even a 9 user. Most don't I agree but why breaking change and drop working platforms when a test will do? React native supports to 9 and consistency has its own value as well. Modern only APIs should be version checked instead of just breaking old ios until the ecosystem moves in my opinion 

@ibrahimsulai thank for checking 😁

**Comment by IbrahimSulai** [comment]
@mikehardy , If there are no more things pending w.r.t this PR, can we get this merged? Thanks.

**Comment by mikehardy** [comment]
Done! Great work thank you!

**Comment by IbrahimSulai** [comment]
Thanks @mikehardy.

**Comment by bd-arc** [comment]
Hi @IbrahimSulai,

Thanks a lot for your work on those issues!

Unless I haven't understood the proper way of sharing images, it seems that there still is a problem with base64 images.

I'm using your update and I've implemented something along the lines of what @xliank does [here](https://github.com/react-native-community/react-native-share/issues/533#issuecomment-508459727). But when I pass my base64 image as the `url` parameter in `Share.shareSingle()`: 

1. for Instagram it opens the App Store page instead of the app (whereas it opens the app if I ditch the `url` parameter)
2. for Facebook it makes the website crash (haven't tested with the app installed yet).

Do you have any idea about what could be going wrong?

**Comment by MateusAndrade** [comment]
@bd-arc you tried using your file path insteatd of passing a base64? Since i started using rn-share, sharing base64 files looks problematic on some cases.

**Comment by bd-arc** [comment]
@MateusAndrade Thank you for the suggestion!

I did actually, but it didn't work either. I used the temporary file path retrieved with `RNFetchBlob.config({ fileCache: true }).then((res: any) => { return res.data; });` to no avail.

Do you know of a better way to retrieve it?

I'm considering downloading the file to the camera roll and then getting the local identifier, as per @itsdante's [suggestion](https://github.com/react-native-community/react-native-share/issues/388#issuecomment-464363636). But this is an overkill that requires specific user permissions and will probably not work on Android...

**Comment by IbrahimSulai** [comment]
@bd-arc - I have created a below sample for passing base64 image directly in shareSingle and its working perfectly for Facebook, Instagram and Twitter.

https://github.com/IbrahimSulai/react-native-share-test/tree/master

In your case - Please try to get the base64 Image from the url separately and give the retrieved base64 Image to the url parameter in the Share.shareSingle().

Please share your sample if you are still not able to resolve your problem. Thanks.

**Comment by bd-arc** [comment]
Thanks a lot @IbrahimSulai for taking the time to put an example together! I really appreciate it 👍

I'm going to use it as a starting point and will get back to you when I've either made it work or put together a test case.

**Comment by bd-arc** [comment]
First of all, thanks again for tackling this issues @IbrahimSulai. Indeed, there wasn't any problem with your code.

I've finally managed to properly share remote images by converting them to base64 images. I'll share some of my findings below since it might prove useful to others too.

---

⚠️ The following is for **iOS only**. Didn't have a chance to try it on Android yet.

---

There were two requirements to make this work that I wasn't aware of and about which I didn't read in any of the issues I browsed:

1. **Your app's `Info.plist` must contain an `NSPhotoLibraryUsageDescription` key**, otherwise it will crash as soon as you try to share your image — on a real device at least, because the simulator stands still. Whether you're downloading the image first, importing a local image64 file or reading a remote URL, you need to include this permission. This wasn't obvious to me, at all.

2. **You need to add every scheme you intend to use in an `LSApplicationQueriesSchemes` key**. This is something I completely overlooked. As an example, here's what I'm using for Instagram and Facebook:

```
<key>LSApplicationQueriesSchemes</key>
<array>
    <string>fb</string>
    <string>fbapi</string>
    <string>fbauth2</string>
    <string>fbshareextension</string>
    <string>fb-messenger-api</string>
    <string>instagram</string>
    <string>instagram-stories</string>
</array>
```

---

And here's how I handle image retrieval/conversion:

```
private share (socialRef?: 'FACEBOOK' | 'INSTAGRAM') {
    RNFetchBlob.fetch('GET', this.props.shareImageUrl)
        .then((res: any) => {
            Share.shareSingle({
                message: 'Your message',
                social: Share.Social[socialRef],
                url:  'data:image/jpeg;base64,' + res.base64(),
                type: 'image/jpeg',
                failOnCancel: false
            })
            .catch(() => {
                // Catch error
            });
        })
        .catch(() => {
            // Catch error
        });
}
```

---

Notes:

* This code is far from perfect since it will create a new image on every call. The proper way to do so would be saving the image to a specific location, checking if it already exists on subsequent calls and loading it from the file system if need be. [`react-native-fs`](https://github.com/itinance/react-native-fs) and [`rn-fetch-blob`](https://github.com/joltup/rn-fetch-blob) are great plugins to help with that.

* If the Facebook app isn't installed, the plugin will default on opening the browser. **With a base64 image, it will actually crash the Facebook website**. I recommend checking whether the app is installed or not thanks to a plugin like [this one](https://github.com/redpandatronicsuk/react-native-check-app-install).

Hope this helps!

**Comment by MateusAndrade** [comment]
@bd-arc thanks for sharing your solution. Would you mind to send a PR updating the README.MD showing this? I see that this is a common issue to people that use react-native-share.

Thanks again for sharing your solution. :smiley:

**Comment by IbrahimSulai** [comment]
@bd-arc @MateusAndrade  - Great to hear that my PR is working as expected. 
Thanks for providing steps to converting the remote image into a base64 image. 
I am always open to take any issues coming in future regarding this functionality. I hope it should not be any :)
Thanks

**Comment by bd-arc** [comment]
Here you are @MateusAndrade: #552 ;-)

File: ios/InstagramShare.m [hunk 1]
```diff
@@ -50,4 +51,74 @@ - (void)shareSingle:(NSDictionary *)options
...
+        [[UIApplication sharedApplication] openURL: [NSURL URLWithString:@"instagram://camera"]];
+        successCallback(@[]);
+    }
+}
+
+-(void)savePictureAndOpenInstagram:(UIImage *)base64Image
+                   failureCallback:(RCTResponseErrorBlock)failureCallback
+                   successCallback:(RCTResponseSenderBlock)successCallback {
+    
+    NSURL *URL = [self fileURLWithTemporaryImageData:UIImageJPEGRepresentation(base64Image, 0.9)];
```
Comment (ProdigySim) [inline ios/InstagramShare.m:85]: Is this recompressing the image before sending it to Instagram?

Comment (jgcmarins) [reply ios/InstagramShare.m:85]: it seems so

**Pull request merged**